File: model/src/main/java/io/github/runtimemodels/chazm/entity/AttributeFactory.java
package io.github.runtimemodels.chazm.entity;

import io.github.runtimemodels.chazm.entity.Attribute.Type;
import io.github.runtimemodels.chazm.id.UniqueId;

/**
 * The {@linkplain AttributeFactory} interface defines the APIs for constructing {@linkplain Attribute}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface AttributeFactory {

    /**
     * Constructs an {@linkplain Attribute}.
     *
     * @param id   the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
     * @param type the {@linkplain Type} of the {@linkplain Attribute}.
     * @return an {@linkplain Attribute}.
     */
    Attribute buildAttribute(UniqueId<Attribute> id, Type type);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/AttributeFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.AttributeId
import runtimemodels.chazm.api.id.UniqueId

/**
 * The [AttributeFactory] interface defines the APIs for constructing [Attribute]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface AttributeFactory {

    /**
     * Constructs an [Attribute].
     *
     * @param id   the [UniqueId] that represents the [Attribute].
     * @param type the [Attribute.Type] of the [Attribute].
     * @return an [Attribute].
     */
    fun build(id: AttributeId, type: Attribute.Type): Attribute

}
